File: src/engine/costs.ts
import type { LayoutVariant, ZoneCategory } from '../domain/types';
import {
  COST_REGIONS,
  OBJECT_COST_TABLE,
  OBJECT_CATEGORY_FALLBACK_COST,
  PATH_COST_PER_M2,
  FENCE_COST_PER_M,
  type CostRegion,
} from '../domain/costData';
import { ZONE_CATEGORY_ORDER, CATEGORY_STYLES } from '../domain/categories';
import { distance } from './geometry';

export interface CostRow {
  label: string;
  installUsd: number;
  annualUsd: number;
}

export interface CostEstimate {
  region: CostRegion;
  landAreaM2: number;
  landCostUsd: number;
  rows: CostRow[];
  constructionTotalUsd: number;
  annualMaintenanceTotalUsd: number;
  totalUpfrontUsd: number;
}

function pathLength(points: { x: number; y: number }[]): number {
  let len = 0;
  for (let i = 0; i < points.length - 1; i++) len += distance(points[i], points[i + 1]);
  return len;
}

export function computeCostEstimate(variant: LayoutVariant, region: CostRegion): CostEstimate {
  const landAreaM2 = variant.analytics.totalAreaM2;
  const landCostUsd = landAreaM2 * region.landPricePerM2Usd;

  const byCategory = new Map<ZoneCategory, { install: number; annual: number }>();
  const addToCategory = (category: ZoneCategory, install: number, annual: number) => {
    const row = byCategory.get(category) ?? { install: 0, annual: 0 };
    row.install += install;
    row.annual += annual;
    byCategory.set(category, row);
  };

  for (const obj of variant.objects) {
    if (obj.locked) continue; // as-built structures aren't a new expense
    const entry = OBJECT_COST_TABLE[obj.typeId] ?? OBJECT_CATEGORY_FALLBACK_COST[obj.category];
    if (!entry) continue;
    const areaM2 = obj.transform.width * obj.transform.height;
    const animalCount = typeof obj.metadata.animalCount === 'number' ? obj.metadata.animalCount : 0;
    const install =
      ((entry.installPerM2 ?? 0) * areaM2 + (entry.installFixed ?? 0)) * region.laborIndex;
    const annual =
      ((entry.annualPerM2 ?? 0) * areaM2 + (entry.annualFixed ?? 0) + (entry.annualPerAnimal ?? 0) * animalCount) *
      region.maintenanceIndex;
    addToCategory(obj.category as ZoneCategory, install, annual);
  }

  let pathInstall = 0;
  let pathAnnual = 0;
  for (const p of variant.paths) {
    const cost = PATH_COST_PER_M2[p.surfaceType] ?? PATH_COST_PER_M2.gravel;
    const areaM2 = pathLength(p.points) * p.widthM;
    pathInstall += cost.install * areaM2 * region.laborIndex;
    pathAnnual += cost.annual * areaM2 * region.maintenanceIndex;
  }

  let fenceInstall = 0;
  let fenceAnnual = 0;
  for (const f of variant.fences) {
    const cost = FENCE_COST_PER_M[f.fenceType] ?? FENCE_COST_PER_M.garden;
    const lengthM = pathLength([...f.points, f.points[0]]);
    fenceInstall += cost.install * lengthM * region.laborIndex;
    fenceAnnual += cost.annual * lengthM * region.maintenanceIndex;
  }

  const rows: CostRow[] = ZONE_CATEGORY_ORDER.filter((c) => byCategory.has(c)).map((category) => {
    const row = byCategory.get(category)!;
    return { label: CATEGORY_STYLES[category].label, installUsd: row.install, annualUsd: row.annual };
  });
  if (pathInstall > 0 || pathAnnual > 0) rows.push({ label: 'Paths', installUsd: pathInstall, annualUsd: pathAnnual });
  if (fenceInstall > 0 || fenceAnnual > 0) rows.push({ label: 'Fencing', installUsd: fenceInstall, annualUsd: fenceAnnual });

  const constructionTotalUsd = rows.reduce((sum, r) => sum + r.installUsd, 0);
  const annualMaintenanceTotalUsd = rows.reduce((sum, r) => sum + r.annualUsd, 0);

  return {
    region,
    landAreaM2,
    landCostUsd,
    rows,
    constructionTotalUsd,
    annualMaintenanceTotalUsd,
    totalUpfrontUsd: landCostUsd + constructionTotalUsd,
  };
}

export function findRegion(id: string): CostRegion {
  return COST_REGIONS.find((r) => r.id === id) ?? COST_REGIONS[0];
}

File: swift/Sources/HomesteadEngine/Costs.swift
import Foundation

/// Ported from `src/engine/costs.ts`. A real bill-of-quantities view turned
/// straight into money — as opposed to `Materials`, which exposes the raw
/// quantities a contractor would actually order against.
public struct CostRow: Equatable, Codable, Sendable {
    public let label: String
    public let category: ZoneCategory?
    public let special: SpecialRow?
    public let installUsd: Double
    public let annualUsd: Double
}

public enum SpecialRow: String, Codable, Sendable {
    case paths, fencing
}

public struct CostEstimate: Equatable, Codable, Sendable {
    public let region: CostRegion
    public let landAreaM2: Double
    public let landCostUsd: Double
    public let rows: [CostRow]
    public let constructionTotalUsd: Double
    public let annualMaintenanceTotalUsd: Double
    public let totalUpfrontUsd: Double
}

public enum Costs {
    private static func polylineLength(_ points: [Point]) -> Double {
        var length = 0.0
        for i in 0..<max(0, points.count - 1) { length += distance(points[i], points[i + 1]) }
        return length
    }

    public static func estimate(objects: [PlanObject], paths: [PathEntity], fences: [Fence], totalAreaM2: Double, region: CostRegion) -> CostEstimate {
        let landAreaM2 = totalAreaM2
        let landCostUsd = landAreaM2 * region.landPricePerM2Usd

        // A JavaScript Map accumulates in first-insertion order; the two
        // running sums below (construction/annual totals) read from this in
        // ZONE_CATEGORY_ORDER, not insertion order, so — unlike Analytics —
        // no tie to the object array's order survives into the output here.
        var byCategory: [ZoneCategory: (install: Double, annual: Double)] = [:]
        func add(_ category: ZoneCategory, install: Double, annual: Double) {
            var entry = byCategory[category] ?? (0, 0)
            entry.install += install
            entry.annual += annual
            byCategory[category] = entry
        }

        for object in objects {
            if object.locked { continue } // as-built structures aren't a new expense
            guard let category = object.category.zoneCategory else { continue }
            guard let entry = CostData.objectCostTable[object.typeId] ?? CostData.objectCategoryFallbackCost[object.category] else { continue }
            let areaM2 = object.transform.width * object.transform.height
            let animalCount = object.metadata["animalCount"]?.doubleValue ?? 0
            let install = ((entry.installPerM2 ?? 0) * areaM2 + (entry.installFixed ?? 0)) * region.laborIndex
            let annual = ((entry.annualPerM2 ?? 0) * areaM2 + (entry.annualFixed ?? 0) + (entry.annualPerAnimal ?? 0) * animalCount) * region.maintenanceIndex
            add(category, install: install, annual: annual)
        }

        var pathInstall = 0.0, pathAnnual = 0.0
        for path in paths {
            let cost = CostData.pathCostPerM2[path.surfaceType] ?? CostData.pathCostPerM2[.gravel]!
            let areaM2 = polylineLength(path.points) * path.widthM
            pathInstall += cost.install * areaM2 * region.laborIndex
            pathAnnual += cost.annual * areaM2 * region.maintenanceIndex
        }

        var fenceInstall = 0.0, fenceAnnual = 0.0
        for fence in fences {
            let cost = CostData.fenceCostPerM[fence.fenceType] ?? CostData.fenceCostPerM[.garden]!
            guard let first = fence.points.first else { continue }
            let lengthM = polylineLength(fence.points + [first])
            fenceInstall += cost.install * lengthM * region.laborIndex
            fenceAnnual += cost.annual * lengthM * region.maintenanceIndex
        }

        var rows: [CostRow] = ZONE_CATEGORY_ORDER.compactMap { category -> CostRow? in
            guard let entry = byCategory[category] else { return nil }
            return CostRow(label: CategoryStyles.label[category]!, category: category, special: nil, installUsd: entry.install, annualUsd: entry.annual)
        }
        if pathInstall > 0 || pathAnnual > 0 {
            rows.append(CostRow(label: "Paths", category: nil, special: .paths, installUsd: pathInstall, annualUsd: pathAnnual))
        }
        if fenceInstall > 0 || fenceAnnual > 0 {
            rows.append(CostRow(label: "Fencing", category: nil, special: .fencing, installUsd: fenceInstall, annualUsd: fenceAnnual))
        }

        let constructionTotalUsd = rows.reduce(0.0) { $0 + $1.installUsd }
        let annualMaintenanceTotalUsd = rows.reduce(0.0) { $0 + $1.annualUsd }

        return CostEstimate(
            region: region,
            landAreaM2: landAreaM2,
            landCostUsd: landCostUsd,
            rows: rows,
            constructionTotalUsd: constructionTotalUsd,
            annualMaintenanceTotalUsd: annualMaintenanceTotalUsd,
            totalUpfrontUsd: landCostUsd + constructionTotalUsd
        )
    }
}

/// The category display label `Costs` borrows from
/// `src/domain/categories.ts`'s `CATEGORY_STYLES[category].label` — the rest
/// of that table (colours) is a view concern with no engine role.
enum CategoryStyles {
    static let label: [ZoneCategory: String] = [
        .residential: "Residential", .access: "Access & Paths", .foodAnnual: "Annual Crops",
        .foodPerennial: "Orchard & Berries", .greenhouse: "Greenhouse", .animal: "Animals",
        .utility: "Utilities", .water: "Water", .energy: "Energy", .storage: "Storage",
        .leisure: "Leisure", .futureExpansion: "Future Expansion",
    ]
}
